h/services/jwt.py

```diff
@@ -1,35 +1,94 @@
-import datetime
+import secrets
+from dataclasses import asdict, dataclass
+from datetime import UTC, datetime, timedelta
+from enum import StrEnum
 from functools import lru_cache
 from typing import Any
 
 import jwt
 from jwt import PyJWKClient
-from jwt.exceptions import PyJWTError
+from jwt.exceptions import InvalidTokenError, PyJWTError
 
 JWK_CLIENT_TIMEOUT = 10
 
 
-class TokenValidationError(Exception):
+class JWTDecodeError(Exception):
     """Decoding or validating a JWT failed."""
 
 
+class JWTPayloadError(JWTDecodeError):
+    """A JWT decoded successfully but the payload wasn't what we expected."""
+
+
+class JWTIssuers(StrEnum):
+    """Strings for use in the `iss` claim when encoding JWTs.
+
+    The actual string values are random so that we can change the enum names
+    without having to worry about the actual string values in JWTs in the wild
+    changing. Anyone decoding a JWT for debugging purposes will see the random
+    strings and can look them up in this enum.
+    """
+
+    OIDC_CONNECT_OR_LOGIN_ORCID = "hak-sax-plix"
+
+
+class JWTAudiences(StrEnum):
+    """Strings for use in the `aud` claim when encoding JWTs.
+
+    The actual string values are random so that we can change the enum names
+    without having to worry about the actual string values in JWTs in the wild
+    changing. Anyone decoding a JWT for debugging purposes will see the random
+    strings and can look them up in this enum.
+    """
+
+    OIDC_REDIRECT_ORCID = "knop-gih-mip"
+
+
+@dataclass
+class OIDCState:
+    """The JWT payload of an OpenID Connect `state` param."""
+
+    action: str
+    """The action that triggered the OIDC flow.
+
+    For example "connect" when connecting an ORCID iD to an existing Hypothesis
+    account, or "login" when using ORCID to log in to Hypothesis
+    """
+
+    rfp: str
+    """A Request Forgery Protection (RFP) token to protect against CSRF attacks.
+
+    The claim name "rfp" is compatible with
+    https://datatracker.ietf.org/doc/html/draft-bradley-oauth-jwt-encoded-state-09
+    """
+
+    @classmethod
+    def make(cls, action):
+        """Return a new OIDCState with the given `action` and a random `rfp`."""
+        return cls(action, secrets.token_hex())
+
+
 class JWTService:
-    LEEWAY = datetime.timedelta(seconds=10)
+    LEEWAY = timedelta(seconds=10)
+
+    def __init__(self, jwt_signing_key):
+        self.jwt_signing_key = jwt_signing_key
 
     @classmethod
-    def decode_token(
+    def decode_oidc_idtoken(
```

Comment:
This existing `decode_token()` method actually implements a particularly decoding algorithm that will only work for some tokens, not others. Renamed it after the one specific use-case that it's actually used for.